pkg/backup/backupdest/backup_destination_test.go

```diff
@@ -337,4 +339,287 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
+	)
+	db.Exec(t, "BACKUP INTO LATEST IN $1", collectionURI)
+	db.Exec(t, "BACKUP INTO LATEST IN $1", collectionURI)
+
+	var aostEnd string
+	db.QueryRow(t, "SELECT now()").Scan(&aostEnd)
+	aostEndTS, err := time.Parse(time.RFC3339, aostEnd)
+	require.NoError(t, err)
+	aostEndHlc := hlc.Timestamp{WallTime: aostEndTS.UnixNano()}
+	db.Exec(
+		t, "BACKUP INTO LATEST IN $1 AS OF SYSTEM TIME $2::STRING",
+		collectionURI, aostEndHlc.AsOfSystemTime(),
+	)
+
+	var fullSubdir string
+	db.QueryRow(
+		t, `SELECT path FROM [SHOW BACKUPS IN $1] ORDER BY path DESC`, collectionURI,
+	).Scan(&fullSubdir)
+	require.NotEmpty(t, fullSubdir)
+
+	var compactionJob jobspb.JobID
+	db.QueryRow(
+		t,
+		`SELECT crdb_internal.backup_compaction(
+				0, $1, $2, $3::DECIMAL, $4::DECIMAL
+			)`,
+		fmt.Sprintf("BACKUP INTO LATEST IN '%s'", collectionURI),
+		fullSubdir,
+		aostFullHlc.AsOfSystemTime(),
+		aostEndHlc.AsOfSystemTime(),
+	).Scan(&compactionJob)
+	jobutils.WaitForJobToSucceed(t, db, compactionJob)
+
+	// Backup chain: f@t=0, i1@t=1, i2@t=2, i3@t=4, c@t=4
+	// where c compacts i1, i2, and i3.
+	const totalBackups = 5
+	const filesPerManifest = 2 // Opening a manifest requires opening the metadata file and checksum
+	for _, tt := range []struct {
+		includeSkipped   bool
+		includeCompacted bool
+		expectedCount    int
+	}{
+		{
+			includeSkipped:   false,
+			includeCompacted: true,
+			expectedCount:    2,
+		},
+		{
+			includeSkipped:   true,
+			includeCompacted: true,
+			expectedCount:    5,
+		},
+		{
+			includeSkipped:   false,
+			includeCompacted: false,
+			expectedCount:    4,
+		},
+		{
+			includeSkipped:   true,
+			includeCompacted: false,
+			expectedCount:    4,
+		},
+	} {
+		t.Run(
+			fmt.Sprintf("includeSkipped=%t,includeCompacted=%t", tt.includeSkipped, tt.includeCompacted),
+			func(t *testing.T) {
+				baseReadersOpened := tc.Servers[0].MustGetSQLCounter("cloud.readers_opened")
+				uris, manifests, locality, memSize, err := indexedResolveBackupManifests(
+					ctx,
+					&mem,
+					[]string{collectionURI},
+					execCfg.DistSQLSrv.ExternalStorageFromURI,
+					fullSubdir,
+					hlc.Timestamp{},
+					nil, /* encryption */
+					nil, /* kms */
+					username.RootUserName(),
+					tt.includeSkipped,
+					tt.includeCompacted,
+				)
+				defer mem.Shrink(ctx, memSize)
+				require.NoError(t, err)
+				require.Len(t, uris, tt.expectedCount)
+				require.Len(t, manifests, tt.expectedCount)
+				require.Len(t, locality, tt.expectedCount)
+
+				// To compute number of manifests opened, we use the following equation:
+				// totalBackups = number of indexes opened
+				// readersOpened - totalBackups = number of files opened to read manifests
+				// (readersOpened - totalBackups) / filesPerManifest = number of manifests opened
+				readersOpened := tc.Servers[0].MustGetSQLCounter("cloud.readers_opened") - baseReadersOpened
+				manifestsOpened := (readersOpened - totalBackups) / filesPerManifest
```

Comment:
this is a neat trick, but i don't quite understand what's going on: why is the following true?
```
 readersOpened - totalBackups = number of files opened to read manifests
```

Reply:
When using the index in `ResolveBackupManifests`, we run through the following flow:

1. [Open the index of every backup in the chain](https://github.com/kev-cao/cockroach/blob/61352b2540ab764122ab51d79aca1bfbef10f415/pkg/backup/backupdest/backup_destination.go#L792-L794) (`totalBackups` number of files opened).
2. Truncate the chain.
3. Then we [just read all the files needed to get the manifests](https://github.com/kev-cao/cockroach/blob/61352b2540ab764122ab51d79aca1bfbef10f415/pkg/backup/backupdest/backup_destination.go#L819-L824) of the remaining backups after truncation.

So if you exclude the indexes read, the only other files read are for manifests. And we read 2 files per manifest, the manifest itself and the checksum.